### ISSUE TITLE
[JENKINS-39231] WinSW: Automatically terminate runaway processes in Windows services

### ISSUE THREAD

**Issue by timja-bot** [timja-org/jenkins-gh-issues-poc-11-04#3980]
<p>In Jenkins projects we have many users complaining that the slave/agent is "already connected", because they have a runaway slave/agent process. It happens when WinSW gets terminated without executing the process shutdown logic (force kill) or when WinSW fails to terminate the process.</p>

<p>As a part of WinSW 2.0, it would be great to create a logic, which...</p>

<ul>
	<li>records PID of the created process to the disc</li>
	<li>performs status check of the previously spawned process upon restart</li>
	<li>terminates the runaway process if required</li>
</ul>

<p>It can be done via WinSW 2 "plugin"<br/>
Issue: <a href="https://github.com/kohsuke/winsw/issues/125" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/kohsuke/winsw/issues/125</a></p>

---
<details><summary><i>Originally reported by <a title="Oleg Nenashev" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-39231" target="_blank">WinSW: Automatically terminate runaway processes in Windows services</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Oleg Nenashev" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-03-13T16:01:07+00:00
<li><b>imported</b>: 2025-11-05
</ul></i>
</details>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mariem_baccar">mariem_baccar</a>:</i>
<p>I have the same issue with the slave "agent is already connected"". you find attached the whole meassage. I think that this issue decreases the performance of jenkins because it consumes a lot CPU. We are waiting for your help!</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p>So there is a pul request for this feature: <a href="https://github.com/kohsuke/winsw/pull/133" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/kohsuke/winsw/pull/133</a><br/>
The pull request also references the release on GitHub with the binary file for evaluation: <a href="https://github.com/kohsuke/winsw/pull/133" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/kohsuke/winsw/pull/133</a></p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mariem_baccar">mariem_baccar</a>:</i>
<ul>
	<li>This problem is related to <a href="https://issues.jenkins.io/browse/JENKINS-39078" title="Config UI in job broken / unusable" class="issue-link" data-issue-key="JENKINS-39078"><del>JENKINS-39078</del></a>: There is a problem in Docker Slave Plugin 1.0.5 (Fix: <a href="https://github.com/jenkinsci/docker-slaves-plugin/commit/451929125fd8ff39c6f84c30476c26cccb912140" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/docker-slaves-plugin/commit/451929125fd8ff39c6f84c30476c26cccb912140</a>). So, you can uninstall this plugin if it's not needed.</li>
	<li>You must be careful. This plugin is also responsible of many other bugs (<a href="https://issues.jenkins.io/browse/JENKINS-39214" title=""Goals and Options" and "Root POM" not written to config.xml" class="issue-link" data-issue-key="JENKINS-39214"><del>JENKINS-39214</del></a>, docker compose fails after scrutation...)</li>
</ul>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mariem_baccar" class="user-hover" rel="mariem_baccar">mariem_baccar</a> Not sure how your comment is related to Windows Service wrapper. Have you modified the wrong issue?</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mariem_baccar">mariem_baccar</a>:</i>
<p>I am modifying the right issue.You will find below more explications for my problem related to slave:<br/>
-Before installing the plugin "Docker Slave v1.0.5", the slave operates correctly without any problems.<br/>
-After installing this plugin, I encountered many problems. One of them is about slave: I always get this message "slave agent is already connected". For more details, you find attached the whole message.<br/>
Finally, I discover that this plugin recently installed is the source of all my new problems in Jenkins 2.19.1 LTS. After uninstalling this one, all problems are resolved.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mariem_baccar" class="user-hover" rel="mariem_baccar">mariem_baccar</a> I'm pretty sure you're modifying the wrong issue. This is a <b>Feature Request</b> to the <b>Windows Service Wrapper</b>, which has nothing to do with the problem you describe. </p>

<p>If you see the "slave agent is already connected" issue only after the Docker Slaves installation, please comment in  <a href="https://issues.jenkins.io/browse/JENKINS-28492" title="The server rejected the connection: *** is already connected to this master. Rejecting this connection." class="issue-link" data-issue-key="JENKINS-28492"><del>JENKINS-28492</del></a>  or create another issue to Docker Slaves plugin</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=krogan">krogan</a>:</i>
<p>Jenkins master is on 2.32.1<br/>
Master and slaves running Win2012<br/>
The symptoms sound very familiar to a problem where we've had a jenkins slave up.... then we reboot the windows server (slave).<br/>
When the server returns and the slave is automatically started, it hangs around for about 30secs then terminates connection which kills our job.<br/>
We've also witnessed the hosting windows service winsw 1.17 (which auto upgrades to 1.18) bombs out but leaves the java process running.<br/>
The java process is still keeping the slave active to the master for an indiscriminate amount of time (anywhere between 20secs to 2hrs) before eventually dying of its own accord, with no fresh jobs sent or interaction with the windows service.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=krogan" class="user-hover" rel="krogan">krogan</a> Smells like another issue. <br/>
Anyway, makes sense to retest after the WinSW 2 integration. You can install the new Windows service wrapper manually</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: Oleg Nenashev<br/>
Path:<br/>
 core/pom.xml<br/>
 core/src/main/resources/windows-service/jenkins-slave.xml<br/>
 core/src/main/resources/windows-service/jenkins.xml<br/>
 war/pom.xml<br/>
<a href="http://jenkins-ci.org/commit/jenkins/e698d1de41d4311bf5f8b1d2c40b591109e696e2" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/e698d1de41d4311bf5f8b1d2c40b591109e696e2</a><br/>
Log:<br/>
  Update Windows Agent Installer to 1.7 and WinSW to 2.0.2 (#2765)</p>

<ol>
	<li>
	<ol>
		<li>
		<ol>
			<li>WinSW changes</li>
		</ol>
		</li>
	</ol>
	</li>
</ol>


<p>The update includes many fixes and improvements, the full list is provided in the <span class="error">&#91;WinSW changelog&#93;</span>(<a href="https://github.com/kohsuke/winsw/blob/master/CHANGELOG.md" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/kohsuke/winsw/blob/master/CHANGELOG.md</a>). There are several issues referenced in Jenkins bugtracker:</p>

<ul>
	<li><a href="https://issues.jenkins.io/browse/JENKINS-22692" title="Jenkins Windows-Slave throwing exception on shutdown causes connection reset issues" class="issue-link" data-issue-key="JENKINS-22692"><del>JENKINS-22692</del></a>(<a href="https://issues.jenkins-ci.org/browse/JENKINS-22692" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-22692</a>) - Connection reset issues when WinSW gets terminated due to the system shutdown</li>
	<li><a href="https://issues.jenkins.io/browse/JENKINS-23487" title="Add support of shared directories mapping to Windows slave services" class="issue-link" data-issue-key="JENKINS-23487"><del>JENKINS-23487</del></a>(<a href="https://issues.jenkins-ci.org/browse/JENKINS-23487)-" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-23487)-</a> Support of shared directories in WinSW</li>
	<li><a href="https://issues.jenkins.io/browse/JENKINS-39231" title="WinSW: Automatically terminate runaway processes in Windows services" class="issue-link" data-issue-key="JENKINS-39231"><del>JENKINS-39231</del></a>(<a href="https://issues.jenkins-ci.org/browse/JENKINS-39231" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-39231</a>) - Enable Runaway Process Killer by default</li>
	<li><a href="https://issues.jenkins.io/browse/JENKINS-39237" title="Add option to update slave.jar in Windows Service Wrapper on startup (when Jenkins runs on HTTPS)" class="issue-link" data-issue-key="JENKINS-39237"><del>JENKINS-39237</del></a>(<a href="https://issues.jenkins-ci.org/browse/JENKINS-39237" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-39237</a>) - Auto-upgrade of JNLP agent versions on the slaves</li>
</ul>


<ol>
	<li>
	<ol>
		<li>
		<ol>
			<li>Windows Agent Installer changes</li>
		</ol>
		</li>
	</ol>
	</li>
</ol>


<ul>
	<li>Adapt the default configurations to pick fixes above</li>
	<li>Slave => Agent renaming where possible</li>
</ul>


<ol>
	<li>
	<ol>
		<li>
		<ol>
			<li>Jenkins core changes</li>
		</ol>
		</li>
	</ol>
	</li>
</ol>


<ul>
	<li>Modify the configuration template, reference advanced options</li>
	<li>Enable Runaway Process Killer by default</li>
</ul>


<ul>
	<li>Update Windows Agent Installer to 1.7</li>
</ul>


<ul>
	<li>Remove the obsolete jenkins-slave.xml file from the core.</li>
</ul>


<p>Now it is within windows-slave-installer</p>

<ul>
	<li>Use the deployed Snapshot for CI</li>
</ul>


<ul>
	<li>Pick the release version of windows-slave-installer-1.7</li>
</ul>
